A couple of proposals for API enhancement for Resource

vaslabs opened the pull request:
Hello,
This is an enhancement for removing the need of the user implementing the close when using an AutoCloseable with the Resource.

I've added a couple of ways, let me know if you have a preference or something else entirely on mind.

Regards

alexandru commented on core/shared/src/main/scala/cats/effect/Resource.scala:
```diff
@@ -73,6 +73,12 @@ sealed abstract class Resource[F[_], A] {
 
 object Resource extends ResourceInstances {
+  def withAutoClosable[F[_], A, C <: AutoCloseable](acquire:F[C])
+             (implicit evidence: Sync[F]): Resource[F, C] =
```
- [ ] rename to `fromAutoClosable`, for consistency with everything else
- [ ] respect formatting — e.g:
  - there should be a standard space after a punctuation mark like `:`
  - the indentation is two spaces, not 11
  - your block style is from C or C#, not Scala or Java 🙂 the opening `{` should not be on its own line
- [ ] we don't use `evidence` for specifying restrictions for `F[_]`, but `F`, as in `F: Sync[F]`
- [ ] remove the `A` type parameter because you're not using it
- [ ] then rename `C` to `A`

alexandru replied:
Two more 🙂

- [ ] add scala docs
- [ ] move the function below, under all other functions, because we want to emphasize the primary data constructor

alexandru commented on core/shared/src/main/scala/cats/effect/syntax/ResourceSyntax.scala:
```diff
@@ -0,0 +1,32 @@
+
+import cats.effect.{Effect, Resource}
+
+trait ResourceSyntax {
```
I would remove the syntax.

I'm against any added syntax and only tolerate it due to our need to work with type classes, however this isn't defining any type class and it's not improving discoverability or readability.

vaslabs commented:
Yes, everything makes sense, thanks will do

vaslabs commented:
Hello,
I addressed all the comments I think. I also added a test and an example code . It's the first time I'm using the microsite, I compiled it and run it locally and saw that it works. Please let me know if I missed anything.
Cheers

rossabaker commented on core/shared/src/main/scala/cats/effect/Resource.scala:
```diff
@@ -108,6 +109,17 @@ object Resource extends ResourceInstances {
     make(fa)(_ => F.unit)
+
+  /**
+    * Creates a resource that will automatically close (e.g. no need to do inputStream.close at the end).
```
All `Resource`s "automatically close".  Maybe "creates a resource that is released by calling `.close()` on the allocated resource".

rossabaker commented on site/src/main/tut/datatypes/resource.md:
````diff
@@ -42,3 +42,18 @@ val r = for {
 r.use { case (a, b) => IO(println(s"Using $a and $b")) }.unsafeRunSync
 ```
+
+If using an AutoCloseable create a resource without the need of dealing with close
````
Maybe "the need to specify how to close"?

alexandru commented on core/shared/src/main/scala/cats/effect/Resource.scala:
```diff
@@ -108,6 +109,17 @@ object Resource extends ResourceInstances {
     make(fa)(_ => F.unit)
+
+  /**
+    * Creates a resource that is released by calling .close() on the allocated resource after use
```
Small nitpick, I'd rather have a ScalaDoc that mentions:

1. Java's `AutoClosable` explicitly, we could even add a link to it
2. a usage sample

Documenting the params or the type params isn't useful on the other hand, if the text doesn't say anything that's not obvious.

```scala
/** Creates a [[Resource]] by wrapping a Java 
   * [[https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html AutoCloseable]].
   *
   * Example:
   * {{{
   *   import java.io._
   * 
   *   def reader[F[_]](file: File): Resource[F, BufferedReader] =
   *     Resource.fromAutoCloseable(F.delay {
   *       new BufferedReader(new FileReader(file))
   *     })
   * }}}
   */
```

There, isn't that more informative? 🙂

vaslabs replied:
it is, to be honest I was feeling a bit lazy for this. I'm of the school that thinks nobody reads the docs but I'll do it that way :)

alexandru replied:
Nobody reads the docs unless you've got awesome docs and I'm proud to say that we're close to having awesome docs 🙂

Daenyth replied:
I read the docs all the time. The best way to get people to not read the docs is to have bad/missing docs. :P

alexandru commented on site/src/main/tut/datatypes/resource.md:
```diff
@@ -42,3 +42,18 @@ val r = for {
+  scala.io.Source.fromString("Hello world")
+}
+
+Resource.fromAutoCloseable(acquire).use(source => IO.pure(println(source.mkString("")))).unsafeRunSync()
```
Here too, wouldn't it be better if the sample actually did something more useful than reading a `"Hello world"` string? 🙂

```scala
import java.io._

def reader[F[_]](file: File)(implicit F: Sync[F]): Resource[F, BufferedReader] =
  Resource.fromAutoCloseable(F.delay {
    new BufferedReader(new FileReader(file))
  })

def dumpResource[F[_]](res: Resource[F, BufferedReader])(implicit F: Sync[F]): F[String] = {
  def loop(in: BufferedReader): F[Unit] =
    F.suspend {
      val line = in.readLine()
      if (line != null) {
        System.out.println(line)
        loop(in)
      } else {
        F.unit
      }
    }
  res.use(loop)
}

def dumpFile[F[_]](file: File)(implicit F: Sync[F]): F[String] =
  dumpResource(reader(file))
```

N.B. we can afford these nitpicks because we've got time until the next release 😜

vaslabs replied:
I don't know, that is a lot of information on the screen. Personally I got a bit dizzy :)
The most common usecase I was trying to make out here is for people I see using the Source and mkString in one liners thinking (including me on my first steps) that the resource is closed at the end.

vaslabs replied:
We could include both though, I think the simple hello world scenario will be more comprehensible for more beginner levels or even a fast read for all people in general. I can add your proposal as a second example?

alexandru replied:
To be honest I have never, ever used `Source` and I don't even know what `Source.fromString` returns.

The `scala.io` package is dead to me.

alexandru replied:
> We could include both though

The more, the merrier.

kubukoz commented on site/src/main/tut/datatypes/resource.md:
```diff
@@ -42,3 +42,67 @@ val r = for {
+import collection.JavaConverters._
+import cats.effect.IO
+
+def readAllLines(bufferedReader: BufferedReader): List[String] = {
```
Should this be suspended in IO?

vaslabs replied:
I realised the whole example was wrong. Fixed

kubukoz commented:
I think this PR could be renamed to mention `AutoCloseable`, as that's the only change here. Also, some conflicts in `Resource.scala`
